File: src/main/java/org/cubexmc/metro/command/newcmd/MetroMainCommand.java
package org.cubexmc.metro.command.newcmd;

import org.incendo.cloud.annotations.CommandDescription;
import org.incendo.cloud.annotations.Command;
import org.incendo.cloud.annotations.Permission;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.cubexmc.metro.Metro;
import org.cubexmc.metro.manager.LineManager;
import org.cubexmc.metro.manager.StopManager;
import org.cubexmc.metro.update.ConfigUpdater;
import org.cubexmc.metro.update.DataFileUpdater;
import org.cubexmc.metro.util.OwnershipUtil;

public class MetroMainCommand {

    private final Metro plugin;
    private final LineManager lineManager;
    private final StopManager stopManager;

    public MetroMainCommand(Metro plugin, LineManager lineManager, StopManager stopManager) {
        this.plugin = plugin;
        this.lineManager = lineManager;
        this.stopManager = stopManager;
    }

    @Command("m|metro")
    @CommandDescription("Metro Main Command")
    public void root(CommandSender sender) {
        help(sender);
    }

    @Command("m|metro help")
    @CommandDescription("Show Metro Help Menu")
    public void help(CommandSender sender) {
        org.cubexmc.metro.manager.LanguageManager lang = plugin.getLanguageManager();
        sender.sendMessage(lang.getMessage("command.help_header"));
        sender.sendMessage(lang.getMessage("command.help_gui"));
        sender.sendMessage(lang.getMessage("command.help_reload"));
        sender.sendMessage(lang.getMessage("command.help_line"));
        sender.sendMessage(lang.getMessage("command.help_stop"));
        sender.sendMessage(lang.getMessage("command.help_portal"));
    }

    @Command("m|metro gui")
    @CommandDescription("Open the Metro GUI")
    @Permission("railway.gui")
    public void gui(Player player) {
        plugin.getGuiManager().openMainMenu(player);
    }

    @Command("m|metro reload")
    @CommandDescription("Reload Metro configuration")
    @Permission("railway.admin")
    public void reload(CommandSender sender) {
        if (sender instanceof Player player) {
            if (!OwnershipUtil.hasAdminBypass(player)) {
                sender.sendMessage(plugin.getLanguageManager().getMessage("plugin.no_permission"));
                return;
            }
        }

        plugin.flushPersistentData();
        plugin.ensureDefaultConfigs();
        plugin.reloadConfig();
        ConfigUpdater.applyDefaults(plugin, "config.yml");
        plugin.getConfigFacade().reload();
        DataFileUpdater.migrateAll(plugin);
        lineManager.reload();
        stopManager.reload();
        if (plugin.getPortalManager() != null) {
            plugin.getPortalManager().load();
        }
        if (plugin.getRailProtectionManager() != null) {
            plugin.getRailProtectionManager().rebuildAll();
        }
        plugin.getLanguageManager().loadLanguages();

        plugin.refreshMapIntegrations();

        sender.sendMessage(plugin.getLanguageManager().getMessage("plugin.reload"));
    }
}

File: src/main/java/org/cubexmc/metro/command/newcmd/MetroMainCommand.kt
package org.cubexmc.metro.command.newcmd

import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import org.cubexmc.config.MigrationException
import org.cubexmc.metro.Metro
import org.cubexmc.metro.manager.LineManager
import org.cubexmc.metro.manager.StopManager
import org.cubexmc.metro.update.DataFileUpdater
import org.cubexmc.metro.update.MetroMigrations
import org.cubexmc.metro.util.OwnershipUtil
import org.incendo.cloud.annotations.Command
import org.incendo.cloud.annotations.CommandDescription
import org.incendo.cloud.annotations.Permission

class MetroMainCommand(
    private val plugin: Metro,
    lineManager: LineManager?,
    stopManager: StopManager?,
) {

    private val lineManagerRef: LineManager? = lineManager
    private val stopManagerRef: StopManager? = stopManager

    private val lineManager: LineManager
        get() = lineManagerRef ?: throw NullPointerException("lineManager")

    private val stopManager: StopManager
        get() = stopManagerRef ?: throw NullPointerException("stopManager")

    @Command("rw|railway|rail")
    @CommandDescription("Railway Main Command")
    fun root(sender: CommandSender) {
        help(sender)
    }

    @Command("rw|railway|rail help")
    @CommandDescription("Show Metro Help Menu")
    fun help(sender: CommandSender) {
        val lang = plugin.languageManager
        sender.sendMessage(lang.getMessage("command.help_header"))
        sender.sendMessage(lang.getMessage("command.help_gui"))
        sender.sendMessage(lang.getMessage("command.help_reload"))
        sender.sendMessage(lang.getMessage("command.help_line"))
        sender.sendMessage(lang.getMessage("command.help_stop"))
        sender.sendMessage(lang.getMessage("command.help_portal"))
    }

    @Command("rw|railway|rail gui")
    @CommandDescription("Open the Metro GUI")
    @Permission("railway.gui")
    fun gui(player: Player) {
        plugin.guiManager.openMainMenu(player)
    }

    @Command("rw|railway|rail reload")
    @CommandDescription("Reload Metro configuration")
    @Permission("railway.admin")
    fun reload(sender: CommandSender) {
        if (sender is Player && !OwnershipUtil.hasAdminBypass(sender)) {
            sender.sendMessage(plugin.languageManager.getMessage("plugin.no_permission"))
            return
        }

        plugin.flushPersistentData()
        plugin.ensureDefaultConfigs()
        try {
            MetroMigrations.migrateConfig(plugin)
            MetroMigrations.ensureEntityDefaults(plugin)
            MetroMigrations.ensureLanguageResources(plugin)
            MetroMigrations.migrateBundledLanguages(plugin)
        } catch (ex: MigrationException) {
            plugin.logger.warning("Railway reload aborted: configuration migration failed: " + ex.message)
            sender.sendMessage("§cRailway reload aborted: configuration migration failed.")
            return
        }
        plugin.reloadConfig()
        plugin.configFacade.reload()
        DataFileUpdater.migrateAll(plugin)
        lineManager.reload()
        stopManager.reload()
        plugin.portalManager?.load()
        plugin.railProtectionManager?.rebuildAll()
        plugin.languageManager.loadLanguages()

        plugin.lineServiceManager?.rebuildFromLines()
        plugin.entityModelController?.reload()

        plugin.refreshMapIntegrations()

        sender.sendMessage(plugin.languageManager.getMessage("plugin.reload"))
    }
}
